Source/JavaScriptCore/runtime/ImportMap.cpp

```diff
@@ -254,18 +284,189 @@ Expected<void, String> ImportMap::registerImportMap(const SourceCode& sourceCode
+                // the developer console.
+                reporter.reportWarning(makeString("An import map scope rule for specifier '"_s, specifier, "' was removed, as it conflicted with already resolved module specifiers."_s));
+                // 3.1.2.1.2. Remove scopeImports[specifier].
+                scopeImports.remove(specifier);
+            }
+        }
+
+        // 3.2 If scopePrefix exists in oldImportMap's scopes, then set
+        // oldImportMap's scopes[scopePrefix] to the result of merging module
+        // specifier maps, given scopeImports and oldImportMap's
+        // scopes[scopePrefix].
+        const auto oldScopeSpecifierMapIt = m_scopesMap.find(scope.key);
+        if (oldScopeSpecifierMapIt != m_scopesMap.end()) {
+            ImportMap::SpecifierMap& oldScopeSpecifierMap = oldScopeSpecifierMapIt->value;
+            ImportMapInternal::mergeModuleSpecifierMaps(oldScopeSpecifierMap, scopeImports, reporter);
+        } else {
+            // 3.3 Otherwise, set oldImportMap's scopes[scopePrefix] to
+            // scopeImports.
+            m_scopesMap.set(scope.key, WTFMove(scopeImports));
+            m_scopesVector.append(scope.key);
+        }
+    }
+
+    // 4. For each url → integrity of newImportMap's integrity:
+    for (auto& url : newImportMapIntegrity.keys()) {
+        const auto& newIntegrityValue = newImportMapIntegrity.get(url);
+        // 4.1 If url exists in oldImportMap's integrity, then:
+        if (m_integrity.contains(url)) {
+            // 4.1.1. The user agent may report the removed rule as a warning to the
+            // developer console.
+            reporter.reportWarning(makeString("An import map integrity rule for url '"_s, url.string(), "' was removed, as it conflicted with already defined integrity rules."_s));
+            // 4.1.2 Continue.
+            continue;
+        }
+        // 4.2 Set oldImportMap's integrity[url] to integrity.
+        m_integrity.set(url, newIntegrityValue);
+    }
+    // 5. For each pair of global's resolved module set:
+
+    // 5.1. For each specifier → url of newImportMapImports:
+
+    // 5.1.1. If specifier starts with pair's specifier, then:
+
+    // We're using a different algorithm here where the resolved module set is
+    // replaced with a set of all the prefixes of specifier resolved. For each
+    // such prefix that exists in the new import map's imports section, we remove
+    // it from that section.
+    for (auto& specifier : m_toplevelResolvedModuleSet) {
+        if (!newImportMapImports.contains(specifier))
+            continue;
+        // 5.1. The user agent may report the removed rule as a warning to the
+        // developer console.
+        reporter.reportWarning(makeString("An import map rule for specifier '"_s, specifier, "' was removed, as it conflicted with already resolved module specifiers."_s));
+        // 5.2. Remove newImportMapImports[specifier].
+        newImportMapImports.remove(specifier);
```

Comment:
Avoid double-lookup. (`newImportMapImports.contains` and `newImportMapImports.remove` are looking up twice). Use `iter` from `find` and use that `iter` for removal.